### PR TITLE
Fix unexpected behavior with `FORMAT` and `SETTINGS` parsing

### LINKED ISSUE #39712: FORMAT and SETTINGS parsing leads to errors
Query parsing related to `FORMAT` and `SETTINGS` clauses is inconsistent.

`select 1 settings log_comment='test' Format {{format}}` works as intended
`(select 1) settings log_comment='test' Format {{format}}` and `(select 1) union (select 2) settings log_comment='test' Format {{format}}` give a parsing error:

```
Syntax error: failed at position 50 ('Format'):

(select 1)  settings log_comment='prova valerio' Format JSONCompact

Expected end of query
```

Queries without the `FORMAT` clause work as intended.

**Does it reproduce on recent release?**
22.7.1.2484 is affected

**How to reproduce**
Connect to clickhouse-client and run the queries

**Expected behavior**

Same behaviour between different versions

> If applicable, add screenshots to help explain your problem.

image

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to write `SETTINGS` before `FORMAT` in a chain of queries with `UNION` when subqueries are inside parentheses. This closes #39712. Change the behavior when a query has the SETTINGS clause specified twice in a sequence. The closest SETTINGS clause will have a preference for the corresponding subquery. In the previous versions, the outermost SETTINGS clause could take a preference over the inner one.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 0d46e7555b066298603b5e0cd4dc122e74863ebf with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60197/0d46e7555b066298603b5e0cd4dc122e74863ebf/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60197/0d46e7555b066298603b5e0cd4dc122e74863ebf/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/InterpreterSetQuery.h [hunk 1]
```diff
@@ -23,7 +23,7 @@ class InterpreterSetQuery : public IInterpreter, WithMutableContext
     /** Set setting for current context (query context).
       * It is used for interpretation of SETTINGS clause in SELECT query.
       */
-    void executeForCurrentContext(bool ignore_setting_constraints = false);
```
Comment (pufit) [inline src/Interpreters/InterpreterSetQuery.h:26]: Is it better without default? I think it's not clear what `executeForCurrentContext(false)` means without looking into the definition (or we can add comments `executeForCurrentContext(/* ignore_setting_constraints = */ false`)

File: src/Parsers/ParserQueryWithOutput.cpp [hunk 2]
```diff
@@ -152,37 +151,55 @@ bool ParserQueryWithOutput::parseImpl(Pos & pos, ASTPtr & node, Expected & expec
...
+      * So while these forms have a slightly different meaning, they both exist:
+      *   SELECT 1 SETTINGS a=b FORMAT TSV
+      *   SELECT 1 FORMAT TSV SETTINGS e=f
+      * And due to this effect, the users expect that the FORMAT and SETTINGS may go in an arbitrary order.
+      * But while this work:
+      *   (SELECT 1) UNION ALL (SELECT 2) FORMAT TSV SETTINGS d=f
+      * This does not work automatically, unless we explicitly allow different orders:
+      *   (SELECT 1) UNION ALL (SELECT 2) SETTINGS d=f FORMAT TSV
+      * Inevitably, we also allow this:
+      *   SELECT 1 SETTINGS a=b SETTINGS d=f FORMAT TSV
```
Comment (pufit) [inline src/Parsers/ParserQueryWithOutput.cpp:175]: This looks funny, but I agree. I don't see an easy solution for this behavior unless we set a strict order in `FORMAT ... SETTINGS`

**Review by pufit — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Needs https://github.com/ClickHouse/ClickHouse/pull/68180.
Needs https://github.com/ClickHouse/ClickHouse/pull/68178

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/issues/68375

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/issues/45160

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/issues/41820

**Comment by alexey-milovidov** [comment]
Needs https://github.com/ClickHouse/ClickHouse/pull/68462.

**Pull request merged**